File: src/test/scala/com/abaddon83/cardsGames/testutils/BurracoGameInitTurnTestFactory.scala
package com.abaddon83.cardsGames.testutils

import com.abaddon83.cardsGames.burracoGames.domainModels.burracoGames.initialised.mazzettos.{MazzettoDeck, MazzettoDecks}
import com.abaddon83.cardsGames.burracoGames.domainModels.burracoGames.initialised.playerInGames.PlayerInGame
import com.abaddon83.cardsGames.burracoGames.domainModels.burracoGames.initialised._
import com.abaddon83.cardsGames.burracoGames.domainModels.burracoGames.waitingPlayers.BurracoGameWaitingPlayers
import com.abaddon83.cardsGames.shares.decks.{Card, Deck}
import com.abaddon83.cardsGames.shares.games.GameIdentity
import com.abaddon83.cardsGames.shares.players.PlayerIdentity

import scala.collection.mutable.ListBuffer

case class BurracoGameInitTurnTestFactory(
                                           private val gameIdentity: GameIdentity,
                                           private val player1: PlayerInGame,
                                           private val player2: PlayerInGame,
                                           private val mazzettoDecks: MazzettoDecks,
                                           private val discardPile: DiscardPile,
                                           private val playerTurn: PlayerIdentity
                                         ) {

  def buildTurnPhaseStart(): BurracoGameInitiatedTurnStart = {
    println(s"discardPile().size: ${discardPile.numCards()}")
    BurracoGameInitiatedTurnStart.build(
      gameIdentity = gameIdentity,
      players = List(player1, player2),
      burracoDeck = burracoDeckBuild(),
      mazzettoDecks = mazzettoDecks,
      discardPile = discardPile,
      playerTurn = playerTurn
    )
  }

  def buildTurnPhaseExecution(): BurracoGameInitiatedTurnExecution = {
    BurracoGameInitiatedTurnExecution.build(
      gameIdentity = gameIdentity,
      players = List(player1, player2),
      burracoDeck = burracoDeckBuild(),
      mazzettoDecks = mazzettoDecks,
      discardPile = discardPile,
      playerTurn = playerTurn
    ).testInvariants()
  }

  def buildTurnPhaseEnd(): BurracoGameInitiatedTurnEnd = {
    BurracoGameInitiatedTurnEnd.build(
      gameIdentity = gameIdentity,
      players = List(player1, player2),
      burracoDeck = burracoDeckBuild(),
      mazzettoDecks = mazzettoDecks,
      discardPile = discardPile,
      playerTurn = playerTurn
    ).testInvariants()
  }

  def buildWaitingPlayer(singlePlayer: Boolean = false): BurracoGameWaitingPlayers = {
    BurracoGameWaitingPlayers(
      gameIdentity = gameIdentity,
      players = {
        if (singlePlayer) List(player1)
        else List(player1, player2)
      }
    )
  }


  def burracoDeckBuild(): BurracoDeck = {
    BurracoDeck(
      ListBuffer.from(
        (Deck.allRanksWithJollyCards() ++ Deck.allRanksWithJollyCards()).drop(
          discardPile.numCards() +
            mazzettoDecks.numCards() +
            player1.totalPlayerCards() +
            player2.totalPlayerCards()
        )
      )
    )
  }

  def setPlayer1(player: PlayerInGame): BurracoGameInitTurnTestFactory = {
    this.copy(
      player1 = player
    )
  }

  def setPlayer2Turn(): BurracoGameInitTurnTestFactory = {
    this.copy(
      playerTurn = player2.playerIdentity
    )
  }

  def setDiscardPile(list: List[Card]): BurracoGameInitTurnTestFactory = {
    this.copy(discardPile = DiscardPile(list))
  }
}

object BurracoGameInitTurnTestFactory {
  def apply(gameIdentity: GameIdentity = GameIdentity(), player1Id: PlayerIdentity = PlayerIdentity(), player2Id: PlayerIdentity = PlayerIdentity()): BurracoGameInitTurnTestFactory = {

    val player1 = PlayerInGame.build(player1Id, Deck.allRanksCards().take(11))
    val player2 = PlayerInGame.build(player2Id, Deck.allRanksCards().take(11))

    val mazzettoDecks = MazzettoDecks.build(List(
      MazzettoDeck.build(Deck.allRanksCards().take(11)),
      MazzettoDeck.build(Deck.allRanksCards().take(11))
    ))
    val discardPile = DiscardPile.apply(List())

    new BurracoGameInitTurnTestFactory(
      gameIdentity = gameIdentity,
      player1 = player1,
      player2 = player2,
      mazzettoDecks = mazzettoDecks,
      discardPile = discardPile,
      playerTurn = player1.playerIdentity
    )
  }
}

File: src/test/kotlin/com/abaddon83/burracoGame/testUtils/BurracoGameInitTurnTestFactory.kt
package com.abaddon83.burracoGame.testUtils

import com.abaddon83.burracoGame.writeModel.models.BurracoDeck
import com.abaddon83.burracoGame.writeModel.models.DiscardPile
import com.abaddon83.burracoGame.writeModel.models.MazzettoDeck
import com.abaddon83.burracoGame.writeModel.models.MazzettoDecks
import com.abaddon83.burracoGame.writeModel.models.burracoGameExecutions.BurracoGameExecutionTurnBeginning
import com.abaddon83.burracoGame.writeModel.models.burracoGameExecutions.BurracoGameExecutionTurnEnd
import com.abaddon83.burracoGame.writeModel.models.burracoGameExecutions.BurracoGameExecutionTurnExecution
import com.abaddon83.burracoGame.writeModel.models.burracoGameExecutions.playerInGames.PlayerInGame
import com.abaddon83.burracoGame.writeModel.models.burracoGameWaitingPlayers.BurracoGameWaitingPlayers
import com.abaddon83.burracoGame.writeModel.models.decks.Card
import com.abaddon83.burracoGame.writeModel.models.decks.ListCardsBuilder
import com.abaddon83.burracoGame.writeModel.models.games.GameIdentity
import com.abaddon83.burracoGame.writeModel.models.players.PlayerIdentity

data class BurracoGameInitTurnTestFactory(
        private val gameIdentity: GameIdentity,
        private val player1: PlayerInGame,
        private val player2: PlayerInGame,
        private val mazzettoDecks: MazzettoDecks,
        private val discardPile: DiscardPile,
        private val playerTurn: PlayerIdentity
) {

    fun buildTurnPhaseStart(): BurracoGameExecutionTurnBeginning {
        return BurracoGameExecutionTurnBeginning.create(
                identity = gameIdentity,
                players = listOf(player1, player2),
                burracoDeck = burracoDeckBuild(),
                mazzettoDecks = mazzettoDecks,
                discardPile = discardPile,
                playerTurn = playerTurn
        )
    }

    fun buildTurnPhaseExecution(): BurracoGameExecutionTurnExecution =
            BurracoGameExecutionTurnExecution.create(
                    identity = gameIdentity,
                    players = listOf(player1, player2),
                    burracoDeck = burracoDeckBuild(),
                    mazzettoDecks = mazzettoDecks,
                    discardPile = discardPile,
                    playerTurn = playerTurn
            )

    fun buildTurnPhaseEnd(): BurracoGameExecutionTurnEnd =
            BurracoGameExecutionTurnEnd.create(
                    identity = gameIdentity,
                    players = listOf(player1, player2),
                    burracoDeck = burracoDeckBuild(),
                    mazzettoDecks = mazzettoDecks,
                    discardPile = discardPile,
                    playerTurn = playerTurn
            )

    fun buildWaitingPlayer(singlePlayer: Boolean = false): BurracoGameWaitingPlayers =
            BurracoGameWaitingPlayers(
                    identity = gameIdentity,
                    players = when (singlePlayer) {
                        true -> listOf(player1)
                        false -> listOf(player1, player2)
                    }, burracoDeck = burracoDeckBuild()
            )

    fun burracoDeckBuild(): BurracoDeck {
        val doubleDeck = ListCardsBuilder.allRanksWithJollyCards().plus(ListCardsBuilder.allRanksWithJollyCards())
        val numCardToDrop = discardPile.numCards() + mazzettoDecks.numCards() + player1.totalPlayerCards() + player2.totalPlayerCards()
        return BurracoDeck.create(doubleDeck.drop(numCardToDrop))
    }

    fun setPlayer1(player: PlayerInGame): BurracoGameInitTurnTestFactory = this.copy(player1 = player)

    fun setPlayer2Turn(): BurracoGameInitTurnTestFactory = this.copy(playerTurn = player2.identity())

    fun setDiscardPile(list: List<Card>): BurracoGameInitTurnTestFactory = this.copy(discardPile = DiscardPile.create(list))

    companion object Factory{
        fun create(gameIdentity: GameIdentity = GameIdentity.create(), player1Id: PlayerIdentity = PlayerIdentity.create(), player2Id: PlayerIdentity = PlayerIdentity.create()): BurracoGameInitTurnTestFactory {

            val player1 = PlayerInGame.create(player1Id, ListCardsBuilder.allRanksCards().take(11))
            val player2 = PlayerInGame.create(player2Id, ListCardsBuilder.allRanksCards().take(11))

            val mazzettoDecks = MazzettoDecks.create(listOf(
                    MazzettoDeck.create(ListCardsBuilder.allRanksCards().take(11)),
                    MazzettoDeck.create(ListCardsBuilder.allRanksCards().take(11))
            ))
            val discardPile = DiscardPile.create(listOf())

            return BurracoGameInitTurnTestFactory(
                    gameIdentity = gameIdentity,
            player1 = player1,
            player2 = player2,
            mazzettoDecks = mazzettoDecks,
            discardPile = discardPile,
            playerTurn = player1.identity()
            )
        }
    }
}
